components/sources/plugin-application/src/lib.rs

```diff
@@ -0,0 +1,687 @@
+            },
+            out_node: ElementReference {
+                source_id: Arc::from(self.source_id.as_str()),
+                element_id: start_node_id.into(),
+            },
+        };
+
+        self.send(SourceChange::Insert { element }).await
+    }
+
+    /// Send a batch of source changes efficiently
+    pub async fn send_batch(&self, changes: Vec<SourceChange>) -> Result<()> {
+        for change in changes {
+            self.send(change).await?;
+        }
+        Ok(())
+    }
+
+    /// Get the source ID that this handle is connected to
+    pub fn source_id(&self) -> &str {
+        &self.source_id
+    }
+}
+
+/// A source that allows applications to programmatically inject events.
+///
+/// This source receives events from an [`ApplicationSourceHandle`] and forwards
+/// them to the Drasi query processing pipeline. It also maintains a record of
+/// all insert events for bootstrap purposes.
+///
+/// # Fields
+///
+/// - `base`: Common source functionality (dispatchers, status, lifecycle)
+/// - `config`: Application source configuration
+/// - `app_rx`: Receiver for events from the handle
+/// - `app_tx`: Sender for creating additional handles
+/// - `bootstrap_data`: Accumulated insert events for bootstrap
+/// - `bootstrap_provider_config`: Optional external bootstrap provider configuration
+pub struct ApplicationSource {
+    /// Base source implementation providing common functionality
+    base: SourceBase,
+    /// Application source configuration
+    config: ApplicationSourceConfig,
+    /// Receiver for events from handles (taken when processing starts)
+    app_rx: Arc<RwLock<Option<mpsc::Receiver<SourceChange>>>>,
+    /// Sender for creating new handles
+    app_tx: mpsc::Sender<SourceChange>,
+    /// Accumulated insert events for internal bootstrap
+    bootstrap_data: Arc<RwLock<Vec<SourceChange>>>,
+    /// Optional external bootstrap provider configuration
+    bootstrap_provider_config: Option<BootstrapProviderConfig>,
+}
+
+impl ApplicationSource {
+    /// Create a new application source and its handle.
+    ///
+    /// The event channel is automatically injected when the source is added
+    /// to DrasiLib via `add_source()`.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique identifier for this source instance
+    /// * `config` - Application source configuration
+    ///
+    /// # Returns
+    ///
+    /// A tuple of `(ApplicationSource, ApplicationSourceHandle)` where the handle
+    /// can be used to send events to the source.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the base source cannot be initialized.
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use drasi_plugin_application::{ApplicationSource, ApplicationSourceConfig};
+    ///
+    /// let config = ApplicationSourceConfig::default();
+    /// let (source, handle) = ApplicationSource::new("my-source", config)?;
+    /// ```
+    pub fn new(
```

Comment:
@agentofreality I am trying to construct an ApplicationSource using this `new` function, however, it returns an `ApplicationSourceHandle`, but I actually need the thing that implement the `Source` trait (ApplicationSource) so that I can add it to my builder.

Reply:
nvm, I figured it out